core/src/main/java/cz/o2/proxima/repository/AttributeValueDescriptor.java

```diff
@@ -0,0 +1,999 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.ToString;
+
+/**
+ * Attribute value descriptor.
+ *
+ * @param <T> attribute type
+ */
+public class AttributeValueDescriptor<T> implements Serializable {
+
+  private static final long serialVersionUID = 1L;
+
+  /** Return {@link Descriptor} for give attribute value */
+  @Getter private final Descriptor<T> schemaDescriptor;
+
+  private AttributeValueDescriptor(Descriptor<T> schemaDescriptor) {
+    this.schemaDescriptor = schemaDescriptor;
+  }
+
+  /**
+   * Create {@link AttributeValueDescriptor} with Primitive type.
+   *
+   * @param descriptor primitive type descriptor
+   * @param <T> attribute type
+   * @return Attribute value descriptor
+   */
+  public static <T> AttributeValueDescriptor<T> of(PrimitiveTypeDescriptor<T> descriptor) {
+    return of(descriptor.toDescriptor());
+  }
+
+  /**
+   * Create {@link AttributeValueDescriptor} with Array type
+   *
+   * @param descriptor Array type descriptor
+   * @param <T> attribute type
+   * @return Attribute value descriptor
+   */
+  public static <T> AttributeValueDescriptor<T> of(ArrayTypeDescriptor<T> descriptor) {
+    return of(descriptor.toDescriptor());
+  }
+
+  /**
+   * Create {@link AttributeValueDescriptor} with Structure type
+   *
+   * @param descriptor Structure type descriptor
+   * @param <T> attribute type
+   * @return attribute value descriptor
+   */
+  public static <T> AttributeValueDescriptor<T> of(StructureTypeDescriptor<T> descriptor) {
+    return of(descriptor.toDescriptor());
+  }
+
+  /**
+   * Create {@link AttributeValueDescriptor} from generic {@link Descriptor}.
+   *
+   * @param descriptor Generic descriptor
+   * @param <T> attribute type
+   * @return attribute value descriptor
+   */
+  public static <T> AttributeValueDescriptor<T> of(Descriptor<T> descriptor) {
+    return new AttributeValueDescriptor<>(descriptor);
+  }
+
+  /** Possible attribute and field types */
+  public enum AttributeValueType {
+    STRUCTURE,
+    ARRAY,
+    ENUM,
+    BYTE,
+    INT,
+    LONG,
+    DOUBLE,
+    FLOAT,
+    STRING,
+    BOOLEAN,
+  }
+
+  /**
+   * Default value factory
+   *
+   * @param <T> attribute type
+   */
+  @FunctionalInterface
+  public interface DefaultValueFactory<T> extends Serializable {
+
+    /** Provide default value */
+    T apply();
+  }
+
+  /**
+   * Interface provides default value for attribute or other structure.
+   *
+   * @param <T> value type
+   */
+  public interface WithDefaultValue<T> {
+
+    /** Return {@code true} when default value can be provided. */
+    default boolean hasDefaultValue() {
+      return getDefaultValue() != null;
+    }
+
+    /** Return default value or null */
+    @Nullable
+    T getDefaultValue();
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
nit: maybe we can leave equals & hashcode to lombok

Reply:
👍